# Guided tutorial
The simplest way to use scigenex is to run the following steps of Seurat R package:
<br>
<ul>
  <li>Create Seurat object</li>
  <li>Quality Control</li>
  <li>Normalization</li>
</ul>
and then use the seurat object as input of SciGeneX. 
<br/>
<br>
SciGeneX assumes that the normalized count matrix is saved in the "data" slot of seurat object and can be loaded with GetAssayData(object, slot="data).
If not or if you didn't use seurat, you can also provide a normalized matrix as input.
<br/>

## Loading normalized count matrix
For this tutorial, we will use scRNA-seq dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available on 10x Genomics. 
This dataset contains 2700 single cells sequenced on the Illumina NextSeq 50. 
You can download it [**here**](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).
<br>

### Common Seurat pipeline
In this step, we will run the pre-processing steps from [common seurat pipeline](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). 
If you already have your pre-processed seurat object or your normalized count matrix, you can skip this step.


```{r , eval=F}
library(Seurat)
# ＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿
# Setup the Seurat object
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
# Read 10x data
pbmc <- Read10X(data.dir = "../pbmc3k/filtered_gene_bc_matrices/hg19")

# Create Seurat object
pbmc_seurat <- CreateSeuratObject(counts = pbmc, project = "pbmc3k",
                                  min.cells = 3, 
                                  min.features = 200)
```

```{r , eval=T, echo = F}
# Load 
suppressWarnings(suppressMessages(library(Seurat, quietly = TRUE)))
pbmc <- Read10X(data.dir = "/mnt/NAS7/Workspace/bavaisj/dbfmcl/dataset_pbmc/filtered_gene_bc_matrices/hg19/")
suppressWarnings(pbmc_seurat <- CreateSeuratObject(counts = pbmc, project = "pbmc3k",
                                                   min.cells = 3, 
                                                   min.features = 200))
pbmc_seurat
```


<br>
<br>
We run here the classical steps of the seurat pipeline. For more information, you can check their website [here](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).
```{r , eval=F}
# ＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿
# Run the classical pipeline of Seurat
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾

# Quality control
pbmc_seurat[["percent.mt"]] <- PercentageFeatureSet(pbmc_seurat,
                                                    pattern = "^MT-")
pbmc_seurat <- subset(pbmc_seurat,
                      subset = percent.mt < 5 & nFeature_RNA > 200)

# Normalizing data
pbmc_seurat <- NormalizeData(pbmc_seurat)

# Identification of highly variable genes
pbmc_seurat <- FindVariableFeatures(pbmc_seurat,
                                    selection.method = "vst",
                                    nfeatures = 2000)

# Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat,
                         features = rownames(pbmc_seurat))

# Perform principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat,
                      features = VariableFeatures(object = pbmc_seurat))

# Cell clustering 
pbmc_seurat <- FindNeighbors(pbmc_seurat, dims = 1:10)
pbmc_seurat <- FindClusters(pbmc_seurat, resolution = 0.5)

# Dimension reduction
pbmc_seurat <- RunUMAP(pbmc_seurat, dims = 1:10)
DimPlot(pbmc_seurat, reduction = "umap")
```


```{r , eval=T, echo=F}
# Quality control
pbmc_seurat[["percent.mt"]] <- PercentageFeatureSet(pbmc_seurat, pattern = "^MT-")
pbmc_seurat <- subset(pbmc_seurat, subset = percent.mt < 5 & nFeature_RNA > 200)

# Normalizing
pbmc_seurat <- NormalizeData(pbmc_seurat)

# Identification of highly variable genes
pbmc_seurat <- FindVariableFeatures(pbmc_seurat, selection.method = "vst", nfeatures = 2000)

# Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat, features = rownames(pbmc_seurat), verbose = FALSE)

# Perform principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat, features = VariableFeatures(object = pbmc_seurat), verbose = FALSE)

# Cell clustering 
pbmc_seurat <- FindNeighbors(pbmc_seurat, dims = 1:10, verbose = FALSE)
pbmc_seurat <- FindClusters(pbmc_seurat, resolution = 0.5, verbose = FALSE)

# Dimension reduction
pbmc_seurat <-suppressWarnings(RunUMAP(pbmc_seurat, dims = 1:10, verbose = FALSE))
DimPlot(pbmc_seurat, reduction = "umap")
```
<br>
<br>





## Extract patterns of co-expressed genes using SciGeneX
In this section, we use the previously generated Seurat object as an input example to run SciGeneX main command. This command run the DBFMCL algorithm (Density-based Filtering and Markov CLuster algorithm) that :
<br>
<ul>
  <li>Identify and extract co-expressed genes</li>
  <li>Define gene cluster related to their respective patterns of co-expression</li>
  <li>Store the result in a ClusterSet object</li>
</ul>

```{r , eval=F}
library(scigenex)

# Run DBFMCL
pbmc_scigenex <- DBFMCL(data = pbmc_seurat,
                        k = 80,
                        inflation = 2,
                        min_cluster_size = 5,
                        av_dot_prod_min = 2)
```

```{r , eval=T, echo=F}
suppressWarnings(suppressMessages(library(scigenex, quietly = TRUE)))

# Run DBFMCL
pbmc_scigenex <- DBFMCL(data = pbmc_seurat,
                        k = 80,
                        inflation = 2,
                        min_cluster_size = 5,
                        av_dot_prod_min = 2)
```

<br>
Clusters of genes are stored in the slot gene_patterns. You can access the gene names from a cluster using the command get_genes().
```{r , eval=T, echo=T}
# Extract gene names from all cluster
genes <- get_genes(pbmc_scigenex, cluster = "all")
head(genes)

# Extract gene names from gene cluster 5
genes_5 <- get_genes(pbmc_scigenex, cluster = 5)
genes_5
```
<br>

## Extract top 10 genes from each patterns of co-expression
You can also used the function top_genes to extract the top genes of each cluster of genes. Genes are ranked regarding their mean correlation values with genes in a same cluster and the top genes of each cluster are stored in the slot top_genes of the ClusterSet object.
```{r , eval=F}
# Find top 10 genes for all gene clusters
pbmc_scigenex <- top_genes(pbmc_scigenex, top = 10)
pbmc_scigenex@top_genes
```

```{r , eval=T, echo=F}
# Find top 10 genes for all gene clusters
pbmc_scigenex <- top_genes(pbmc_scigenex, top = 10)
pbmc_scigenex@top_genes
```
<br>
<br>
You can acess the top genes of a specific cluster of genes using the get_genes function.
```{r , eval=F}
# Extract gene names from the top 10 genes of cluster 5
genes_cl5_top <- get_genes(pbmc_scigenex, cluster = 5, top = TRUE)
genes_cl5_top
```

```{r , eval=T, echo=F}
# Extract gene names from the top 10 genes of cluster 5
genes_cl5_top <- get_genes(pbmc_scigenex, cluster = 5, top = TRUE)
genes_cl5_top
```
<br>
<br>
<br>
<br>

## Use SciGeneX as feature selection method
At the end of this step, you can use the genes found by SciGeneX as feature selection results. In the followed example, those genes are stored in the Seurat object and we re-run the Seurat pipeline using those genes.


```{r , eval=F}
# ＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿＿
# Run the classical pipeline of Seurat
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾

# Use genes found by SciGeneX as feature selection results
pbmc_seurat@assays$RNA@var.features <- get_genes(pbmc_scigenex)

# Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat,
                         features = rownames(pbmc_seurat))

# Perform principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat,
                      features = VariableFeatures(object = pbmc_seurat))

# Cell clustering 
pbmc_seurat <- FindNeighbors(pbmc_seurat, dims = 1:10)
pbmc_seurat <- FindClusters(pbmc_seurat, resolution = 0.5)

# Dimension reduction
pbmc_seurat <- RunUMAP(pbmc_seurat, dims = 1:10)
DimPlot(pbmc_seurat, reduction = "umap")
```


```{r , eval=T, echo=F}
# Identification of highly variable genes
pbmc_seurat@assays$RNA@var.features <- get_genes(pbmc_scigenex)

# Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat, features = rownames(pbmc_seurat), verbose = FALSE)

# Perform principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat, features = VariableFeatures(object = pbmc_seurat), verbose = FALSE)

# Cell clustering 
pbmc_seurat <- FindNeighbors(pbmc_seurat, dims = 1:10, verbose = FALSE)
pbmc_seurat <- FindClusters(pbmc_seurat, resolution = 0.5, verbose = FALSE)

# Dimension reduction
pbmc_seurat <-suppressWarnings(RunUMAP(pbmc_seurat, dims = 1:10, verbose = FALSE))
DimPlot(pbmc_seurat, reduction = "umap")
```
<br>
<br>

## Heatmap visualization
```{r , eval=T, echo=T}
cell_cluster <- pbmc_seurat[["seurat_clusters"]]
cell_cluster <- rownames(cell_cluster)[order(cell_cluster[,"seurat_clusters"])]

plot_heatmap(pbmc_scigenex,
             cell_order = cell_cluster,
             use_top_genes = TRUE,
             line_size = 2) %>%
  add_col_annotation(pbmc_seurat[["seurat_clusters"]][cell_cluster,"seurat_clusters"]) %>%
  add_col_barplot(colMeans(GetAssayData(pbmc_seurat, slot = "data")[,cell_cluster]))
```
<br>
<br>

## Investigate genes in co-expression patterns
### Cell proliferation
```{r , eval=F}
# Functional enrichment analysis
pbmc_scigenex <- enrich_go(pbmc_scigenex, specie = "Hsapiens", ontology = "BP")
pbmc_scigenex <- viz_enrich(pbmc_scigenex, type = "barplot", nb_terms = 5)
pbmc_scigenex@cluster_annotations$plot_cl5$barplot
```

```{r , eval=T, echo=F}
# Functional enrichment analysis
pbmc_scigenex <- enrich_go(pbmc_scigenex, specie = "Hsapiens", ontology = "BP")
pbmc_scigenex <- viz_enrich(pbmc_scigenex, type = "barplot", nb_terms = 5)
pbmc_scigenex@cluster_annotations$plot_cl5$barplot
```

```{r , eval=T, echo=T}
# Visualization of the top gene expression in UMAP
col_greyMagma = c("2"="grey", "4"="#FB8861FF", "5"="#B63679FF", "3"="#51127CFF", "1"="#000004FF")
FeaturePlot(pbmc_seurat,
            features = get_genes(pbmc_scigenex,cluster = 5, top = TRUE)[1:4],
            cols = col_greyMagma,
            order = TRUE)


# Compute module score using the gene list of a specific patterns of co-expression
pbmc_seurat <- AddModuleScore(pbmc_seurat,
                             features = list(get_genes(pbmc_scigenex, cluster = 5)),
                             name = "Cells_in_proliferation")

FeaturePlot(pbmc_seurat,
            features = "Cells_in_proliferation1",
            order = TRUE,
            cols = col_greyMagma)

```



### B cells
```{r , eval=T, echo=T}
# Functional enrichment analysis
pbmc_scigenex@cluster_annotations$plot_cl6$barplot


# Visualization of the top gene expression in UMAP
FeaturePlot(pbmc_seurat,
            features = get_genes(pbmc_scigenex,cluster = 6, top = TRUE)[1:4],
            cols = col_greyMagma,
            order = TRUE)


# Compute module score using the gene list 
# of a specific patterns of co-expression
pbmc_seurat <- AddModuleScore(pbmc_seurat,
                             features = list(get_genes(pbmc_scigenex, cluster = 6)),
                             name = "B_cells")

FeaturePlot(pbmc_seurat,
            features = "B_cells1",
            order = TRUE,
            cols = col_greyMagma)

```
<br>
<br>





