# Guided tutorial
The simplest way to use scigenex is to run the following steps of Seurat R package:
<br>
<ul>
  <li>Create Seurat object</li>
  <li>Quality Control</li>
  <li>Normalization</li>
</ul>
and then use the seurat object as input of SciGeneX. 
<br/>
<br>
SciGeneX assumes that the normalized count matrix is saved in the "data" slot of seurat object and can be loaded with GetAssayData(object, slot="data).
If not or if you didn't use seurat, you can also provide a normalized matrix as input.
<br/>

## Loading normalized count matrix
For this tutorial, we will use scRNA-seq dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available on 10x Genomics. 
This dataset contains 2700 single cells sequenced on the Illumina NextSeq 50. 
You can download it [**here**](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).
<br>

### Common Seurat pipeline
In this step, we will run the pre-processing steps from [common seurat pipeline](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). 
If you already have your pre-processed seurat object or your normalized count matrix, you can skip this step.

#### Setup the Seurat object
```{r , eval=F}
library(Seurat)

# Read 10x data
pbmc <- Read10X(data.dir = "../pbmc3k/filtered_gene_bc_matrices/hg19")

# Create Seurat object
pbmc_seurat <- CreateSeuratObject(counts = pbmc, project = "pbmc3k",
                                  min.cells = 3, 
                                  min.features = 200)
```

```{r , eval=T, echo = F}
# Load 
suppressMessages(library(Seurat, quietly = TRUE))
suppressMessages(library(scigenex, quietly = TRUE))
pbmc <- Read10X(data.dir = "/mnt/NAS7/Workspace/bavaisj/dbfmcl/dataset_pbmc/filtered_gene_bc_matrices/hg19/")
suppressWarnings(pbmc_seurat <- CreateSeuratObject(counts = pbmc, project = "pbmc3k",
                                                   min.cells = 3, 
                                                   min.features = 200))
pbmc_seurat
```


#### Classical pipeline of Seurat
We run here the classical steps of the seurat pipeline. For more information, you can check their website [here](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).
```{r , eval=F}
# Quality control
pbmc_seurat[["percent.mt"]] <- PercentageFeatureSet(pbmc_seurat, pattern = "^MT-")
pbmc_seurat <- subset(pbmc_seurat, subset = percent.mt < 5 & nFeature_RNA > 200)

# Normalizing
pbmc_seurat <- NormalizeData(pbmc_seurat)

# Identification of highly variable genes
pbmc_seurat <- FindVariableFeatures(pbmc_seurat, selection.method = "vst", nfeatures = 2000)

# Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat, features = rownames(pbmc_seurat))

# Perform principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat, features = VariableFeatures(object = pbmc_seurat))

# Cell clustering 
pbmc_seurat <- FindNeighbors(pbmc_seurat, dims = 1:10)
pbmc_seurat <- FindClusters(pbmc_seurat, resolution = 0.5)

# Dimension reduction
pbmc_seurat <- RunUMAP(pbmc_seurat, dims = 1:10)
DimPlot(pbmc_seurat, reduction = "umap")
```


```{r , eval=T, echo=F}
# Quality control
pbmc_seurat[["percent.mt"]] <- PercentageFeatureSet(pbmc_seurat, pattern = "^MT-")
pbmc_seurat <- subset(pbmc_seurat, subset = percent.mt < 5 & nFeature_RNA > 200)

# Normalizing
pbmc_seurat <- NormalizeData(pbmc_seurat)

# Identification of highly variable genes
pbmc_seurat <- FindVariableFeatures(pbmc_seurat, selection.method = "vst", nfeatures = 2000)

# Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat, features = rownames(pbmc_seurat), verbose = FALSE)

# Perform principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat, features = VariableFeatures(object = pbmc_seurat), verbose = FALSE)

# Cell clustering 
pbmc_seurat <- FindNeighbors(pbmc_seurat, dims = 1:10, verbose = FALSE)
pbmc_seurat <- FindClusters(pbmc_seurat, resolution = 0.5, verbose = FALSE)

# Dimension reduction
pbmc_seurat <-suppressWarnings(RunUMAP(pbmc_seurat, dims = 1:10, verbose = FALSE))
DimPlot(pbmc_seurat, reduction = "umap")
```
<br>
<br>





## Extract patterns of co-expressed genes using SciGeneX
In this section, we use the previously generated Seurat object as an input example to run SciGeneX main command. This command run the DBFMCL algorithm (Density-based Filtering and Markov CLuster algorithm) that :
<br>
<ul>
  <li>Identify and extract co-expressed genes</li>
  <li>Define gene cluster related to their respective patterns of co-expression</li>
  <li>Store the result in a ClusterSet object</li>
</ul>

```{r , eval=T, echo=T}
# Run DBFMCL
pbmc_scigenex <- DBFMCL(data = pbmc_seurat,
                        k = 80,
                        inflation = 2,
                        min_cluster_size = 5,
                        av_dot_prod_min = 2)
```

The gene clusters are stored in the slot gene_patterns. You can access the gene names from a cluster using the command get_genes().
```{r , eval=T, echo=T}
# Extract gene names from all cluster
genes <- get_genes(pbmc_scigenex, cluster = "all")
head(genes)

# Extract gene names from gene cluster 5
genes_5 <- get_genes(pbmc_scigenex, cluster = 5)
genes_5
```
<br>

## Extract top 10 genes from each patterns of co-expression
```{r , eval=F}
# Find top 10 genes for all gene clusters
pbmc_scigenex <- top_genes(pbmc_scigenex, top = 10)
pbmc_scigenex@top_genes

# Extract gene names from the top 10 genes of cluster 5
genes_5_top <- get_genes(pbmc_scigenex, cluster = 5, top = TRUE)
genes_5_top
```

```{r , eval=T, echo=F}
# Find top 10 genes for all gene clusters
pbmc_scigenex <- top_genes(pbmc_scigenex, top = 10)
pbmc_scigenex@top_genes

# Extract gene names from the top 10 genes of cluster 5
genes_5_top <- get_genes(pbmc_scigenex, cluster = 5, top = TRUE)
genes_5_top
```
<br>
<br>

## Cell clustering
```{r , eval=F}
# Perform hierarchical cluster algorithm and cut the tree using DynamicTreeCut package
pbmc_scigenex <- cell_clust(pbmc_scigenex, min_cluster_size = 4)
```

```{r , eval=T, echo=F}
# Perform hierarchical cluster algorithm and cut the tree using DynamicTreeCut package
pbmc_scigenex <- cell_clust(pbmc_scigenex, min_cluster_size = 4)
```
<br>

## Heatmap visualization
We provide a command to generate an interactive heatmap.
```{r , eval=T, echo=T}
# Heatmap of the top genes for each pattern of co-expressed genes
plot_heatmap(pbmc_scigenex,
             use_top_genes = TRUE,
             cell_order = names(sort(pbmc_scigenex@cell_clusters$labels)),
             line_size = 2)

# Using core cells of each cell clusters
plot_heatmap(pbmc_scigenex,
             use_top_genes = TRUE,
             use_core_cells = TRUE,
             line_size = 2)
```
<br>

## Gene cluster enrichment analysis
```{r , eval=F}
# Enrichment analysis
pbmc_scigenex <- enrich_go(pbmc_scigenex, specie = "Hsapiens")
pbmc_scigenex <- viz_enrich(pbmc_scigenex, type = "barplot", nb_terms = 25)
pbmc_scigenex@cluster_annotations$plot_cl5$barplot
```

```{r , eval=T, echo=F, fig.height=20, fig.height=16}
# Enrichment analysis
pbmc_scigenex <- enrich_go(pbmc_scigenex, specie = "Hsapiens")
pbmc_scigenex <- viz_enrich(pbmc_scigenex, type = "barplot", nb_terms = 25)
pbmc_scigenex@cluster_annotations$plot_cl5$barplot
```
<br>
<br>

## Example of identification of rare cell population
```{r , eval=T, echo=T}
# Visualization of a co-expressed gene pattern expressed in a small number of genes
plot_heatmap(pbmc_scigenex,
             use_top_genes = TRUE,
             use_core_cells = TRUE,
             line_size = 2,
             cluster = 7)

# In the top genes of the pattern, we identify genes that are known to be marker of plasmacytoid dendritic cells (LILRA4, CLEC4C, SERPINF1)
get_genes(pbmc_scigenex, cluster = 7, top = TRUE)


# Mapping the cell cluster on UMAP
## Use genes selected by SciGeneX as feature selection

## Scaling data
pbmc_seurat <- ScaleData(pbmc_seurat, features = rownames(pbmc_seurat))
## Principal component analysis
pbmc_seurat <- RunPCA(pbmc_seurat, features = VariableFeatures(object = pbmc_seurat))
## Dimension reduction
pbmc_seurat <- RunUMAP(pbmc_seurat, dims = 1:10)


DimPlot(pbmc_seurat, reduction = "umap", cells.highlight = get_cells(pbmc_scigenex, cluster = 13))

```








