---
author: "Julie Bavais"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{Guided tutorial}
  %\usepackage[UTF-8]{inputenc}
---

# Guided tutorial

The simplest way to use scigenex is to run the following steps using the Seurat R package:

* Load data into a Seurat object
* Perform quality Control
* Perform normalization


The subsequent object can be used as input to SciGeneX. Alternatively, you can provide a normalized matrix as input.


## The dataset

For this tutorial, we will use scRNA-seq dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available on 10x Genomics web site. This dataset contains 2700 single cells sequenced on the Illumina NextSeq 500.  This dataset can be downloaded [**from 10X Genomics**](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz) or through the `SeuratData`library

### Preparing the pbmc3k dataset

In this step, we will run the classical pre-processing steps from the. Please refer to [this tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) for more informations. 
If you have already pre-processed your data with Seurat or have a normalized count matrix as input, you can skip this step.


```{r loading1, echo=FALSE, eval=TRUE, warning=FALSE, results='hide'}
suppressMessages(suppressWarnings(library(ggplot2, quietly = TRUE)))
suppressMessages(suppressWarnings(library(patchwork, quietly = TRUE)))
suppressMessages(suppressWarnings(library(Seurat, quietly = TRUE)))
suppressMessages(suppressWarnings(library(SeuratData, quietly = TRUE)))
suppressMessages(suppressWarnings(InstallData("pbmc3k")))
suppressMessages(suppressWarnings(data("pbmc3k")))

```

```{r loading2, echo=TRUE, eval=FALSE}
library(ggplot2)
library(patchwork)
library(Seurat, quietly = TRUE)
library(SeuratData)

InstallData("pbmc3k")
data("pbmc3k")

```

We next run the classical steps of the seurat pipeline. For more information, you can check Seurat website [here](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).

```{r classicalsteps, echo=TRUE, eval=TRUE, warning=FALSE}
# Quality control
pbmc3k[["percent.mt"]] <- PercentageFeatureSet(pbmc3k, pattern = "^MT-")
pbmc3k <- subset(pbmc3k, subset = percent.mt < 5 & nFeature_RNA > 200)

# Normalizing
pbmc3k <- NormalizeData(pbmc3k)

# Identification of highly variable genes
pbmc3k <- FindVariableFeatures(pbmc3k, selection.method = "vst", nfeatures = 2000)

# Scaling data
pbmc3k <- ScaleData(pbmc3k, features = rownames(pbmc3k), verbose = FALSE)

# Perform principal component analysis
pbmc3k <- RunPCA(pbmc3k, features = VariableFeatures(object = pbmc3k), verbose = FALSE)

# Cell clustering
pbmc3k <- FindNeighbors(pbmc3k, dims = 1:10, verbose = FALSE)
pbmc3k <- FindClusters(pbmc3k, resolution = 0.5, verbose = FALSE)

# Dimension reduction
pbmc3k <-suppressWarnings(RunUMAP(pbmc3k, dims = 1:10, verbose = FALSE))
DimPlot(pbmc3k, reduction = "umap")
```

## Extracting gene modules using SciGeneX

In this section, we use the previously generated Seurat object as an input to run SciGeneX main commands. This command run the main algorithm that will:
 
* Identify and extracts co-expressed genes
* Partition selected genes into clusters
* Store the result in a ClusterSet object

First we will load the Scigenex library. To limit the verbosity of Scigenex function we will set the verbosity level to zero (which will switch off information messages and debugging messages).


```{r loadlib1, warning=FALSE, results='hide', echo = FALSE, eval=TRUE}
suppressMessages(suppressWarnings(library(scigenex)))
scigenex::set_verbosity(0)
```

```{r loadlib2, echo = TRUE, eval=FALSE, warning=FALSE, results='hide'}
library(scigenex)
scigenex::set_verbosity(0)
```

Then we call successively the `select_genes()` function which will select informative genes (*i.e* co-regulated), then the `gene_clustering()` function which will call MCL and partition the dataset into gene modules.

```{r scigenex, echo=TRUE, eval=TRUE, results='hide', warning=FALSE }

# Select informative genes
pbmc_scigenex <- select_genes(pbmc3k,
                              k = 50,
                              distance_method = "pearson",
                              which_slot = "data",
                              row_sum=2)

# Run MCL
pbmc_scigenex <- gene_clustering(pbmc_scigenex,
                                 s = 5,
                                 threads = 2,
                                 inflation = 2)
```

The object produced is a `ClusterSet` objet that is a S4 object that is intented to store gene modules. 

```{r}
isS4(pbmc_scigenex)
pbmc_scigenex
```

There are various methods associated with the `ClusterSet` objects.


```{r}
show_methods()
```

The current object contains `r nrow(pbmc_scigenex)` informative genes, `r ncol(pbmc_scigenex)` samples and `r nclust(pbmc_scigenex)` gene modules.

```{r}
nrow(pbmc_scigenex)
ncol(pbmc_scigenex)
nclust(pbmc_scigenex)
```

At this step, several modules need to be filtered out as lots of them may be singletons. Interestingly, the `ClusterSet` class implements the indexing operator/function ("["). The first argument/dimension of the indexing function corresponds to the cluster stored in the object. The second dimension correspond to cell names. As an example, one can simply keep gene modules with size (*i.e* number of genes) greater than 7 using the following code. This leads to an object containing `r nclust(pbmc_scigenex[clust_size(pbmc_scigenex) > 7, ])` gene modules.

```{r}
pbmc_scigenex <- pbmc_scigenex[clust_size(pbmc_scigenex) > 7, ]
nclust(pbmc_scigenex)
```

It may also be important to filter out gene based on dispersion. Several parameters can be computed for each cluster using the `cluster_stats()`function.

```{r}
plot_cluster_stats(cluster_stats(pbmc_scigenex)) + 
 ggplot2::theme(axis.text.y = ggplot2::element_blank(),
                axis.text.x = element_text(angle=45, vjust = 0.5),
                panel.grid = element_blank()) 

```

Here we will select gene modules based on standard deviation (> 0.1) and rename the cluster (from 1 to the number of clusters):

```{r}
pbmc_scigenex <- pbmc_scigenex[cluster_stats(pbmc_scigenex)$sd_total > 0.1, ] 
pbmc_scigenex <- rename_clust(pbmc_scigenex)
nclust(pbmc_scigenex)
```

Then we check the statistics again.

```{r}
plot_cluster_stats(cluster_stats(pbmc_scigenex)) + 
 ggplot2::theme(axis.text.y = ggplot2::element_blank(),
                axis.text.x = element_text(angle=45, vjust = 0.5),
                panel.grid = element_blank()) 
```

Clusters of genes are stored in the `gene_clusters` slot. One can access the gene names from a cluster using the `get_genes()` command. By default, all genes are returned.

```{r }
# Extract gene names from the 5th gene cluster
genes_module_5 <- get_genes(pbmc_scigenex, cluster = 5)
head(genes_module_5)
```
One can also access the gene to cluster mapping using `get_genes()`.

```{r}
head(gene_cluster(pbmc_scigenex))
tail(gene_cluster(pbmc_scigenex))
```


## Visualisation using heatmap

### Visualization of specific gene modules

The visualization of a heatmap containing all cells and modules may be time consuming and may require important memory ressources. A first alternative is to look at gene cluster individually or a subset of clusters. The gene modules can be visualized using the `plot_heatmap()` or the `plot_ggheatmap()` functions. The first is mainly intented to propose interactive vizualisation based on the `iheatmapr` library and allows to easily browse the results and zoom on particular regions of the heatmap The second should be mostly used for non interactive vizualisation. This second solution leads to a plot that is more easilly customable as it is bas    ed on the ggplot framework. Here we use `plot_ggheatmap()` to have a look at the 4 first clusters. Note that here, we choose to order columns/cells on `Seurat::FindClusters` results.


```{r}
plot_ggheatmap(pbmc_scigenex[1:4, ], 
               use_top_genes = FALSE,
               ident=Idents(pbmc3k)) + ggtitle("Cluster 1 to 4") 
```

### Heatmap of representative genes

However one alternative is to extract the most representative genes of each cluster. This can be achieved using the `top_genes()` function. This function will store the identifiers of these representative genes inside the `top_genes` slot of the `ClusterSet` object. The `top_genes`slot can be accessed the `get_genes()` function. 

```{r}
pbmc_scigenex <- top_genes(pbmc_scigenex)
genes_cl5_top <- get_genes(pbmc_scigenex, cluster = 5, top = TRUE)
genes_cl5_top
```

Both `plot_heatmap()` and `plot_ggheatmap()` support the `use_top_genes`argument:

```{r, warning=FALSE }

plot_ggheatmap(pbmc_scigenex, 
               use_top_genes = TRUE,
               ident=Seurat::Idents(pbmc3k)) + ggtitle("All clusters (top genes)") +
               theme(strip.text.y = element_text(size=4))
```

### Interactive heatmap

A very interesting feature of Scigenex is its ability to display gene expression levels across cells through interactive heatmaps. Using this feature, user can interactively assess the expression levels in selected cells or cluster or on the whole dataset. However, it is generally advisable, when using all the clusters to restrict analysis using `top_genes=TRUE`. Here we will display the expression levels across cells of cluster 3 to 5.

```{r}
plot_heatmap(pbmc_scigenex[3:5, ], 
               use_top_genes = TRUE,
               cell_clusters =Seurat::Idents(pbmc3k))
```


## Exporting modules

Gene modules can be exported using the `cluster_set_to_xls()`. This function will create a Excel workbook that will contain the mapping from genes to modules.

```{r echo=FALSE}
tmp_file <- file.path(tempdir(), "pbmc_scigenex_out.xls")
out <- suppressWarnings(file.remove(tmp_file))
```

```{r}
tmp_file <- file.path(tempdir(), "pbmc_scigenex_out.xls")
cluster_set_to_xls(object=pbmc_scigenex, file_path=tmp_file)

```

## Functional enrichment analysis

Functional enrichment analysis can be performed for each gene module using the `enrich_go()` function. Enrichments can be displayed using the `plot_clust_enrichments()` function.

```{r , eval=TRUE, echo=TRUE}
# Functional enrichment analysis
pbmc_scigenex <- enrich_go(pbmc_scigenex, specie = "Hsapiens", ontology = "BP")
plot_clust_enrichments(pbmc_scigenex, gradient_palette=colors_for_gradient("Je1"), 
                       floor=50,
                       nb_go_term = 2) + 
  theme(axis.text.y = element_text(size=4))
```

## Mapping cell populations markers onto the gene modules

Given a set of markers, the `plot_markers_to_clusters()` function can be used to map cell type markers to gene modules. This function will display jaccard and hypergeometric statistics. Here we will use the markers from the [sctype.app](https://sctype.app/) database.

```{r eval=FALSE, echo=TRUE}
library(dplyr)
library(tidyr)
```

```{r eval=TRUE, echo=FALSE}
suppressMessages(suppressWarnings(library(dplyr)))
suppressMessages(suppressWarnings(library(tidyr)))
```

```{r}
sctype <- "https://zenodo.org/record/8269433/files/sctype.app.tsv"
marker_table <- read.table(sctype, head=TRUE, sep="\t")
marker_table <- marker_table %>% 
                filter(Tissue == "Immune system") %>% 
                filter(!grepl('Pro-|Pre-|HSC|precursor|Progenitor', Cell.type)) %>%
                separate_rows(Marker_genes, convert = TRUE)

markers <- split(marker_table$Marker_genes, 
                 marker_table$Cell.type)
plot_markers_to_clusters(pbmc_scigenex, 
                         markers=markers, background = rownames(pbmc3k))
```


## Working with spatial transcriptomics datasets

The scigenex package offers a certain number of function dedicated to spatial transcriptomic data analysis. At the moment these functions have been mainly developed to analyse VISIUM technology (10X Genomics).

### Searching for gene modules

### Pre-process spatial transcriptomics data

Here we will use the *stxBrain* dataset as example. This dataset is available from SeuratData library and contains mouse brain spatial expression over in several datasets. Two datasets are for the posterior region, two for the anterior. We will use the *anterior1* dataset that we will first pre-process using Seurat. 

```{r}
brain1 <- LoadData("stxBrain", type = "anterior1")
brain1 <- NormalizeData(brain1, 
                        normalization.method = "LogNormalize",
                        verbose = FALSE)
brain1 <- ScaleData(brain1, verbose = FALSE)
brain1 <- FindVariableFeatures(brain1, verbose = FALSE)
brain1 <- RunPCA(brain1, assay = "Spatial", verbose = FALSE)
brain1 <- FindNeighbors(brain1, reduction = "pca", dims = 1:20, verbose = FALSE)
brain1 <- FindClusters(brain1, verbose = FALSE)
brain1 <- RunUMAP(brain1, reduction = "pca", dims = 1:20, verbose = FALSE)

DimPlot(brain1, reduction = "umap", label = TRUE)
SpatialDimPlot(brain1, label = TRUE, label.size = 3, pt.size.factor = 1.4)
```

### Calling scigenex

Here we will used the alternative clustering methods proposed by scigenex ("reciprocal_neighborhood"). It is advise to increase slightly k (here k will be set to 80). After call to gene_clustering(). apply filtering on gene modules based on cluster size (min number of genes 7) and standard deviation (gene module sd > 0.1). 

```{r spatialrun, echo=TRUE, eval=TRUE, warning=FALSE}
res_brain <- select_genes(data=brain1,
                    k=80,
                    distance_method="pearson",
                    row_sum = 5)

gc_brain <- gene_clustering(res_brain, 
                            method = "reciprocal_neighborhood", 
                            inflation = 2, 
                            threads = 4)
gcs_brain <- filter_cluster_size(gc_brain, min_cluster_size = 7)
df <- cluster_stats(gcs_brain) 
gcss_brain <- gcs_brain[df$sd_total > 0.1, ]
gcss_brain <- rename_clust(gcss_brain)
length(row_names(gcss_brain))
nclust(gcss_brain)
```

Interestingly, Scigenex algorithm is able to retrieve `nclust(gcss_brain)` gene modules. This is most probably reminiscent of cell complexity but also of numerous molecular pathways that are differentially activated across the organ and unanticipated complexity.  

### Visualizing corresponding heatmap

We then may display the corresponding heatmap using `plot_ggheatmap()`.

```{r heatspatial, warning=FALSE }
gcss_brain <- top_genes(gcss_brain)
plot_ggheatmap(gcss_brain, 
               use_top_genes = TRUE,
               ident=Idents(brain1)) + ggtitle("All clusters (top genes)") +
               theme(strip.text.y = element_text(size=3))
```

### Interactive heatmap

Again, as in the context of scRNA-seq, we may also  use the powerful `plot_heatmap()` fonction which allows interactive exploration of all or specific clusters. Here we look at cluster 6 to 9.

```{r}
plot_heatmap(gcss_brain[6:9,], 
             use_top_genes = TRUE, 
             cell_clusters = Seurat::Idents(brain1))
# Try selecting a subset of columns/rows
# The 'home' button can be used to reset
# the heatmap
```

### Visualizing topological clusters

The scigenex library implements the `plot_spatial()` function to display topological information. In addition to the signal, specific regions can be highlighted using a hull that can be created using the `display_hull()` function. Here will also add a hull around seurat cluster 0 and 2. We will then display signal for "seurat_clusters" metadata.

```{r hull_white, fig.width=6, fig.height=4}
hull_white <- display_hull(getFlippedTissueCoordinates(brain1),
                           ident = ifelse(Idents(brain1) %in% c(0, 2), 
                                          1, 0),
                           color = "white", 
                           size_x=4, size_y=3, 
                           hull_type = "wall", 
                           size = 0.5, 
                           step_x = 2.6, 
                           step_y=2.4, 
                           delta = 1.5)

plot_spatial(seurat_obj = brain1, 
             metadata = "seurat_clusters", 
             pt_size=2.5, coord_flip = T) + hull_white

```

We may also want to visualize a specific gene. Here we will look at the pattern of "Hpca" which is part of the cluster `r setNames(which_clust( gcss_brain, "Hpca"), NULL)` detected by Scigenex.

```{r}
"Hpca" %in% gcss_brain
Hpca_clust <- which_clust( gcss_brain, "Hpca")
Hpca_clust
```

To this aim we will use the `plot_spatial()` function.

```{r plot_spatial, fig.width=6, fig.height=4}
plot_spatial(seurat_obj = brain1, 
             gene_name = "Hpca",  
             pt_size=2.5, coord_flip = T) + hull_white  +
  ggtitle("Hpca expression pattern.")
```

Note that cluster `r setNames(which_clust( gcss_brain, "Hpca"), NULL)` also contains several genes related to *Regulator Of G Protein* family and *Potassium Channel Tetramerization Domain*. This can be checked with a regular expression using the `grep_clust()` function:

```{r}
grep_clust(gcss_brain[Hpca_clust, ], "(Rgs)|(Kctd)",)
```

To visualize the topological distribution of signals in each cluster we will *(i)* extract the `gene_clusters` slot from the `gcss` object, *(ii)* compute module score using the Seurat `AddModuleScore()`function and *(iii)* store the results in the seurat object. Note that, for each gene module, signal will be scaled from 0 to 1. This will allow us to have a shared legend between all plots

```{r AddModuleScore, echo=TRUE, eval=TRUE, warning=FALSE}
brain1 <- AddModuleScore(brain1, features = gcss_brain@gene_clusters)

for(i in 1:nclust(gcss_brain)){ # Normalizing module scores
  tmp <- brain1[[paste0("Cluster", i, sep="")]] 
  max_tmp <- max(tmp)
  min_tmp <- min(tmp)
  brain1[[paste0("Cluster", i, sep="")]]  <- (tmp[,1] - min(tmp))/(max_tmp - min_tmp)
}
```

The topological profile of cluster `r Hpca_clust` (that contains Hpca, a strong hippocampus marker) is the following:


```{r plot_spatialHpca_clust, fig.width=6, fig.height=4}
plot_spatial(seurat_obj = brain1, 
             metadata = paste("Cluster", setNames(Hpca_clust, NULL), 
                                         sep=""), 
             pt_size=2.5, coord_flip = T) + hull_white
```

We can easily see that the pattern of the gene module containing Hpca is very similar to the Hpca pattern. Moreover this gene module pattern also strongly overlaps the Seurat cell cluster number 0. However, the analysis seems to indicate that a more complex tissue architecture exists as Hpca signal extends beyond Seurat cluster 0.

We will then display the topological signal of all clusters using the `plot_spatial_panel()` function. 

```{r topoclusters, echo=TRUE, eval=TRUE, warning=FALSE, fig.width=6, fig.height=10}
p <- plot_spatial_panel(brain1, 
                        metadata = paste("Cluster", 1:nclust(gcss_brain), 
                                         sep=""), 
                   ncol_layout = 3, 
                   pt_size=0.7, 
                   guide='collect',
                   stroke = 0, size_title = 5, 
                   face_title = 'plain', 
                   barwidth = 0.25, barheight = 1.5, 
                   coord_flip=T) 
print(p + guide_area())
```

## Comparing FindAllMarkers() and Scigenex gene modules

It can be interesting to compare the clusters obtained using unsupervised approach (scigenex) to the list of genes obtained using `Seurat::FindAllMarkers()` that search for differentially expressed genes between the cell populations infered by `Seurat::FindClusters()`. Here the comparison is performed using results obtained for spatial transcriptomics data.

First note that `Seurat::FindAllMarkers()` will tend to find shared markers between populations. 

```{r seurat_classes, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE, results='hide'}
seurat_brain_mk <- Seurat::FindAllMarkers(brain1)
seurat_brain_mk<- split(seurat_brain_mk$gene, seurat_brain_mk$cluster)
plot_cmp_genesets(seurat_brain_mk, seurat_brain_mk, layout = "square", transform="-log10" ) + 
  xlab("Seurat::FindAllMarkers()") + 
  ylab("Seurat::FindAllMarkers")
```

In contrast, gene partitioning in Scigenex is a hard clustering method (elements are not shared between clusters).

```{r}
plot_cmp_genesets(gcss_brain@gene_clusters, gcss_brain@gene_clusters, layout = "square", transform="-log10") + 
  xlab("Scigenex") + 
  ylab("Scigenex") 
```

When comparing both one can see that, although some overlaps exist between gene clusters obtained from both appraoches, they are capturing different information that are probably complementary.

```{r}
plot_cmp_genesets(gcss_brain@gene_clusters, seurat_brain_mk, layout = "square", transform="-log10" )  +
  ylab("Seurat::FindAllMarkers()") + 
  xlab("Scigenex")
```

## Creating a report

A report can be created using the `cluster_set_report()`function. The arguments of these function are the processed clusterSet and the corresponding Seurat processed object. Ideally the clusterSet object should contain functionnal annotation (see `enrich_go()`). Currently, the process of creating a report may be rather long. It may also produce heavy html files that may take some time to load into de web browser.

```{r report, echo=TRUE, eval=FALSE, warning=FALSE, results='hide'}
# Uncomment to prepare the report.
# gcss_brain <- enrich_go(gcss_brain, species = "Mmusculus", ontology = "BP")
# cluster_set_report(clusterset_object = gcss_brain, seurat_object = brain1)
```
An example report is available [here]().


## Session info

```{r}
sessionInfo()
```

