# Guided tutorial
The simplest way to use scigenex is to run the following steps of Seurat R package:
<br>
<ul>
  <li>Create Seurat object</li>
  <li>Quality Control</li>
  <li>Normalization</li>
</ul>
and then use the seurat object as input of SciGeneX. 
<br/>
<br>
SciGeneX assumes that the normalized count matrix is saved in the "data" slot of seurat object and can be loaded with GetAssayData(object, slot="data).
If not or if you didn't use seurat, you can also provide a normalized matrix as input.
<br/>

## Loading normalized count matrix
For this tutorial, we will use scRNA-seq dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available on 10x Genomics. 
This dataset contains 2700 single cells sequenced on the Illumina NextSeq 50. 
You can download it [**here**](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).
<br>

### Common Seurat pipeline
In this step, we will run the pre-processing steps from [common seurat pipeline](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). 
If you already have your pre-processed seurat object or your normalized count matrix, you can skip this step.

#### Setup the Seurat object
```{r , eval=T, include = F}
# Load 
library(Seurat)
pbmc <- Read10X(data.dir = "/mnt/NAS7/Workspace/bavaisj/dbfmcl/dataset_pbmc/filtered_gene_bc_matrices/hg19/")
pbmc_seurat <- CreateSeuratObject(counts = pbmc, project = "pbmc3k",
                                  min.cells = 3, 
                                  min.features = 200)
pbmc_seurat
```

```{r , eval=F, echo=T}
library(Seurat)

# Read 10x data
pbmc <- Read10X(data.dir = "../pbmc3k/filtered_gene_bc_matrices/hg19")

# Create Seurat object
pbmc_seurat <- CreateSeuratObject(counts = pbmc, project = "pbmc3k",
                                  min.cells = 3, 
                                  min.features = 200)
```

#### Quality Control
We filter cells that have less than 5% of mitochondrial genes or 200 unique feature counts or more than 2500 feature counts.
```{r , eval=F, echo=T}
# Compute percentage of mitochondrial genes
pbmc_seurat[["percent.mt"]] <- PercentageFeatureSet(pbmc_seurat, pattern = "^MT-")
# Filter cells
pbmc_seurat <- subset(pbmc_seurat, subset = percent.mt < 5 & nFeature_RNA > 200 & nFeature_RNA < 2500)
```

#### Normalizing count matrix
```{r , eval=F, echo=T}
# Normalizing
pbmc_seurat <- NormalizeData(pbmc_seurat)
```
<br>
<br>





## Extract patterns of co-expressed genes using SciGeneX
In this section, we use the previously generated Seurat object as an input example to run SciGeneX main command. This command run the DBFMCL algorithm (Density-based Filtering and Markov CLuster algorithm) that :
<br>
<ul>
  <li>Identify and extract co-expressed genes</li>
  <li>Define gene cluster related to their respective patterns of co-expression</li>
  <li>Store the result in a ClusterSet object</li>
</ul>

```{r , eval=F, echo=T}
# Run DBFMCL
pbmc_scigenex <- DBFMCL(data = pbmc_seurat,
                        k = 50,
                        inflation = 8,
                        min_cluster_size = 5,
                        av_dot_prod_min=2)
```

The gene clusters are stored in the slot gene_patterns. You can access the gene names from a cluster using the command get_genes().
```{r , eval=F, echo=T}
# Extract gene names from all cluster
genes <- get_genes(pbmc_scigenex, cluster = "all")
genes

# Extract gene names from gene cluster 3
genes_3 <- get_genes(pbmc_scigenex, cluster = 3)
genes_3
```


## Extract top 10 genes from each patterns of co-expression
```{r , eval=F, echo=T}
# Find top 10 genes for all gene clusters
pbmc_scigenex <- top_genes(pbmc_scigenex, top = 10)
pbmc_scigenex@top_genes

# Extract gene names from the top 10 genes of cluster 3
genes_3_top <- get_genes(pbmc_scigenex, cluster = 3, top = TRUE)
genes_3_top
```


## Visualization of gene clusters
We provide a command to generate an interactive heatmap of the gene cluster
```{r , eval=F, echo=T}
# 
plot_heatmap(pbmc_scigenex, use_top_genes = TRUE, line_size = 2)
```



## Gene cluster enrichment analysis
```{r , eval=F, echo=T}
# 
pbmc_scigenex <- enrich_go(pbmc_scigenex, specie = "Hsapiens")
pbmc_scigenex <- viz_enrich(pbmc_scigenex, type = "dotplot")
pbmc_scigenex@cluster_annotations$plot_cl3$barplot
```


## Cell clustering
```{r , eval=F, echo=T}
# 
pbmc_scigenex <- cell_clust(pbmc_scigenex)
plot_heatmap(pbmc_scigenex, use_top_genes = TRUE, line_size = 2)
```










