<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>The "Density Based Filtering and Markov CLustering" algorithm (DBF-MCL). — DBFMCL • scigenex</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="The " density based filtering and markov clustering algorithm dbfmcl><meta property="og:description" content='DBFMCL is a tree-steps adaptative algorithm that (i) find elements
located in dense areas (DBF), (ii) uses selected items to construct a
graph, (iii) performs graph partitioning using the Markov CLustering
Algorithm (MCL).
This function requires installation of the mcl program
(http://www.micans.org/mcl). See "Warnings" section for more
informations.
When analyzing a noisy dataset, one is interested in isolating dense regions
as they are populated with genes/elements that display weak distances to
their nearest neighbors (i.e. strong profile similarities). To isolate these
regions DBF-MCL computes, for each gene/element, the distance with its kth
nearest neighbor (DKNN).In order to define a critical DKNN value that will
depend on the dataset and below which a gene/element will be considered as
falling in a dense area, DBF-MCL computes simulated DKNN values by using an
empirical randomization procedure. Given a dataset containing n genes and p
samples, a simulated DKNN value is obtained by sampling n distance values
from the gene-gene distance matrix D and by extracting the kth-smallest
value. This procedure is repeated n times to obtain a set of simulated DKNN
values S. Computed distributions of simulated DKNN are used to compute a FDR
value for each observed DKNN value. The critical value of DKNN is the one
for which a user-defined FDR value (typically 10%) is observed. Genes with
DKNN value below this threshold are selected and used to construct a graph.
In this graph, edges are constructed between two genes (nodes) if one of
them belongs to the k-nearest neighbors of the other. Edges are weighted
based on the respective coefficient of correlation (i.e., similarity)
and the graph obtained is partitioned using the Markov CLustering Algorithm
(MCL).'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scigenex</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Install</a>
</li>
<li>
  <a href="../articles/usage.html">Guided tutorial</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dputhier/scigenex/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The "Density Based Filtering and Markov CLustering" algorithm (DBF-MCL).</h1>
    <small class="dont-index">Source: <a href="https://github.com/dputhier/scigenex/blob/HEAD/R/dbfmcl_main.R" class="external-link"><code>R/dbfmcl_main.R</code></a></small>
    <div class="hidden name"><code>DBFMCL.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>DBFMCL is a tree-steps adaptative algorithm that <em>(i)</em> find elements
located in dense areas (DBF), <em>(ii)</em> uses selected items to construct a
graph, <em>(iii)</em> performs graph partitioning using the Markov CLustering
Algorithm (MCL).</p>
<p>This function requires installation of the mcl program
(<a href="http://www.micans.org/mcl" class="external-link">http://www.micans.org/mcl</a>). See "Warnings" section for more
informations.</p>
<p>When analyzing a noisy dataset, one is interested in isolating dense regions
as they are populated with genes/elements that display weak distances to
their nearest neighbors (i.e. strong profile similarities). To isolate these
regions DBF-MCL computes, for each gene/element, the distance with its kth
nearest neighbor (DKNN).In order to define a critical DKNN value that will
depend on the dataset and below which a gene/element will be considered as
falling in a dense area, DBF-MCL computes simulated DKNN values by using an
empirical randomization procedure. Given a dataset containing n genes and p
samples, a simulated DKNN value is obtained by sampling n distance values
from the gene-gene distance matrix D and by extracting the kth-smallest
value. This procedure is repeated n times to obtain a set of simulated DKNN
values S. Computed distributions of simulated DKNN are used to compute a FDR
value for each observed DKNN value. The critical value of DKNN is the one
for which a user-defined FDR value (typically 10%) is observed. Genes with
DKNN value below this threshold are selected and used to construct a graph.
In this graph, edges are constructed between two genes (nodes) if one of
them belongs to the k-nearest neighbors of the other. Edges are weighted
based on the respective coefficient of correlation (<em>i.e.</em>, similarity)
and the graph obtained is partitioned using the Markov CLustering Algorithm
(MCL).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">DBFMCL</span><span class="op">(</span>
  data <span class="op">=</span> <span class="cn">NULL</span>,
  filename <span class="op">=</span> <span class="cn">NULL</span>,
  path <span class="op">=</span> <span class="st">"."</span>,
  output_path <span class="op">=</span> <span class="st">"."</span>,
  mcl_threads <span class="op">=</span> <span class="fl">1</span>,
  name <span class="op">=</span> <span class="cn">NULL</span>,
  distance_method <span class="op">=</span> <span class="st">"pearson"</span>,
  av_dot_prod_min <span class="op">=</span> <span class="fl">2</span>,
  min_cluster_size <span class="op">=</span> <span class="fl">10</span>,
  silent <span class="op">=</span> <span class="cn">FALSE</span>,
  k <span class="op">=</span> <span class="fl">50</span>,
  fdr <span class="op">=</span> <span class="fl">10</span>,
  inflation <span class="op">=</span> <span class="fl">8</span>,
  seed <span class="op">=</span> <span class="fl">123</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>a <code>matrix</code>, <code>data.frame</code> or <code>Seurat</code> object.</p></dd>
<dt>filename</dt>
<dd><p>a character string representing the file name.</p></dd>
<dt>path</dt>
<dd><p>a character string representing the data directory where
intermediary files are to be stored. Default to current working directory.</p></dd>
<dt>output_path</dt>
<dd><p>a character string representing the data directory where
output files will be stored. Default to current working directory.</p></dd>
<dt>mcl_threads</dt>
<dd><p>An integer to determine number of threads for mcl algorithm.</p></dd>
<dt>name</dt>
<dd><p>a prefix for the names of the intermediary files created by DBF
and MCL.</p></dd>
<dt>distance_method</dt>
<dd><p>a method to compute the distance to the k-th nearest neighbor.</p></dd>
<dt>av_dot_prod_min</dt>
<dd><p>Any cluster with average dot product below this value is discarded. This allow to delete
clusters in which correlation is influenced/supported by very few samples (typically 1).</p></dd>
<dt>min_cluster_size</dt>
<dd><p>Minimum number of element inside a cluster. MCL tend to create lots of clusters with
very few (e.g 2) objects.</p></dd>
<dt>silent</dt>
<dd><p>if set to TRUE, the progression of distance matrix calculation
is not displayed.</p></dd>
<dt>k</dt>
<dd><p>the neighborhood size.</p></dd>
<dt>fdr</dt>
<dd><p>an integer value corresponding to the false discovery rate
(range: 0 to 100).</p></dd>
<dt>inflation</dt>
<dd><p>the main control of MCL. Inflation affects cluster
granularity. It is usually chosen somewhere in the range <code>[1.2-5.0]</code>.
<code>inflation = 5.0</code> will tend to result in fine-grained clusterings, and
whereas <code>inflation = 1.2</code> will tend to result in very coarse grained
clusterings. By default, <code>inflation = 2.0</code>. Default setting gives very
good results for microarray data when k is set between 70 and 250.</p></dd>
<dt>seed</dt>
<dd><p>specify seeds for random number generator.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a ClusterSets class object.</p>
    </div>
    <div id="warnings">
    <h2>Warnings</h2>
    <p>With the current implementation, this function only works
only on UNIX-like plateforms.</p>
<p>MCL should be installed. One can used the following command lines in a
terminal:</p>
<p><code># Download the latest version of mcl (the script has been tested
successfully with the 06-058 version).</code>
<code>wget http://micans.org/mcl/src/mcl-latest.tar.gz</code>
<code># Uncompress and install mcl</code>
<code>tar xvfz mcl-latest.tar.gz</code>
<code>cd mcl-xx-xxx</code>
<code>./configure</code>
<code>make</code>
<code>sudo make install</code>
<code># You should get mcl in your path</code>
<code>mcl -h</code></p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>- Van Dongen S. (2000) A cluster algorithm for graphs. National
Research Institute for Mathematics and Computer Science in the 1386-3681.
- Lopez F.,Textoris J., Bergon A., Didier G., Remy E., Granjeaud
S., Imbert J. , Nguyen C. and Puthier D. TranscriptomeBrowser: a powerful
and flexible toolbox to explore productively the transcriptional landscape
of the Gene Expression Omnibus database. PLoSONE, 2008;3(12):e4001.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Bergon A., Bavais J., Textoris J., Granjeaud S., Lopez F and Puthier
D.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">  <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">80000</span><span class="op">)</span>, nc<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">m</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op">+</span> <span class="fl">4</span></span>
<span class="r-in">   <span class="va">m</span><span class="op">[</span><span class="fl">101</span><span class="op">:</span><span class="fl">200</span>,<span class="fl">11</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">[</span><span class="fl">101</span><span class="op">:</span><span class="fl">200</span>,<span class="fl">11</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span> <span class="op">+</span> <span class="fl">3</span></span>
<span class="r-in">   <span class="va">m</span><span class="op">[</span><span class="fl">201</span><span class="op">:</span><span class="fl">300</span>,<span class="fl">5</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">[</span><span class="fl">201</span><span class="op">:</span><span class="fl">300</span>,<span class="fl">5</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span> <span class="op">+</span> <span class="op">-</span><span class="fl">2</span></span>
<span class="r-in">   <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">DBFMCL</span><span class="op">(</span>data<span class="op">=</span><span class="va">m</span>,</span>
<span class="r-in">                 distance_method<span class="op">=</span><span class="st">"pearson"</span>,</span>
<span class="r-in">                 av_dot_prod_min <span class="op">=</span> <span class="fl">0</span>,</span>
<span class="r-in">                 inflation <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span class="r-in">                 k<span class="op">=</span><span class="fl">25</span>,</span>
<span class="r-in">                 fdr <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plot_clust.html">plot_clust</a></span><span class="op">(</span><span class="va">res</span>, ceil <span class="op">=</span> <span class="fl">10</span>, floor <span class="op">=</span> <span class="op">-</span><span class="fl">10</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plot_clust.html">plot_clust</a></span><span class="op">(</span><span class="va">res</span>, type<span class="op">=</span><span class="st">"tile"</span>, ceil <span class="op">=</span> <span class="fl">10</span>, floor <span class="op">=</span> <span class="op">-</span><span class="fl">10</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="write_clust.html">write_clust</a></span><span class="op">(</span><span class="va">res</span>, filename_out <span class="op">=</span> <span class="st">"ALL.sign.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by J. Bavais.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

